File: Mobile/lib/services/chat_signalr_service.dart
import 'package:signalr_netcore/signalr_client.dart';
import 'package:flutter/foundation.dart';
import '../http/lib_http.dart';

class ChatSignalRService {
  late HubConnection _hub;

  final List<Function(int, String, DateTime)> _onMessageHandlers = [];
  final List<Function(bool)> _onPartnerOnlineHandlers = [];
  final List<Function()> _onPartnerTypingHandlers = [];
  final List<Function(int, String, String, DateTime)> _onNewConversationHandlers = [];
  final List<Function()> _onTypingClearedHandlers = [];
  final List<Function(int, Map<String, dynamic>, DateTime)> _onIncidentHandlers = [];

  Future<void> connect() async {
    debugPrint('[ChatSignalR] Connecting to $API_URL/hubs/chat');

    _hub = HubConnectionBuilder()
        .withUrl(
      '$API_URL/hubs/chat',
      options: HttpConnectionOptions(
        accessTokenFactory: () async => AuthStore.token ?? '',
        transport: HttpTransportType.WebSockets,
        skipNegotiation: true,
      ),
    )
        .withAutomaticReconnect(retryDelays: [0, 2000, 5000, 10000, 30000])
        .build();

    _hub.serverTimeoutInMilliseconds = 60000;
    _hub.keepAliveIntervalInMilliseconds = 15000;

    _hub.onclose(({error}) {
      debugPrint('[ChatSignalR] Connection closed: $error');
    });

    _hub.onreconnecting(({error}) {
      debugPrint('[ChatSignalR] Reconnecting: $error');
    });

    _hub.onreconnected(({connectionId}) {
      debugPrint('[ChatSignalR] Reconnected: $connectionId');
    });

    _hub.off('ReceiveMessage');
    _hub.on('ReceiveMessage', (args) {
      if (args == null || args.length < 3) return;
      final fromCitizenId = args[0] as int;
      final message = args[1] as String;
      final sentAt = DateTime.parse(args[2] as String);
      for (final h in _onMessageHandlers) h(fromCitizenId, message, sentAt);
      _clearTyping();
    });

    _hub.off('ReceiveIncident');
    _hub.on('ReceiveIncident', (args) {
      if (args == null || args.length < 3) return;
      final fromCitizenId = args[0] as int;
      final incident = Map<String, dynamic>.from(args[1] as Map);
      final sentAt = DateTime.parse(args[2] as String);
      for (final h in _onIncidentHandlers) h(fromCitizenId, incident, sentAt);
      _clearTyping();
    });

    _hub.off('PartnerOnline');
    _hub.on('PartnerOnline', (args) {
      for (final h in _onPartnerOnlineHandlers) h(true);
    });

    _hub.off('PartnerOffline');
    _hub.on('PartnerOffline', (args) {
      for (final h in _onPartnerOnlineHandlers) h(false);
    });

    _hub.off('PartnerTyping');
    _hub.on('PartnerTyping', (args) {
      for (final h in _onPartnerTypingHandlers) h();
    });

    _hub.off('NewConversationMessage');
    _hub.on('NewConversationMessage', (args) {
      if (args == null || args.length < 4) return;
      final fromCitizenId = args[0] as int;
      final fromName = args[1] as String;
      final message = args[2] as String;
      final sentAt = DateTime.parse(args[3] as String);
      for (final h in _onNewConversationHandlers) {
        h(fromCitizenId, fromName, message, sentAt);
      }
    });

    await _hub.start();
    debugPrint('[ChatSignalR] Connected successfully! State: ${_hub.state}');
  }

  Future<List<Map<String, dynamic>>> getChatIncidents() async {
    final res = await SingletonDio.getDio().get('$API_URL/api/Chat/incidents');
    return List<Map<String, dynamic>>.from(res.data);
  }

  Future<void> openConversation(int partnerCitizenId) async {
    await _hub.invoke('OpenConversation', args: [partnerCitizenId]);
  }

  Future<void> closeConversation(int partnerCitizenId) async {
    await _hub.invoke('CloseConversation', args: [partnerCitizenId]);
  }

  Future<void> sendMessage(int toCitizenId, String message) async {
    await _hub.invoke('SendMessage', args: [toCitizenId, message]);
  }

  Future<void> sendIncident(int toCitizenId, int incidentId) async {
    await _hub.invoke('SendIncident', args: [toCitizenId, incidentId]);
  }

  Future<void> sendTyping(int toCitizenId) async {
    await _hub.invoke('Typing', args: [toCitizenId]);
  }

  void onTypingCleared(Function() handler) =>
      _onTypingClearedHandlers.add(handler);

  void _clearTyping() {
    for (final h in _onTypingClearedHandlers) h();
  }

  void onIncident(Function(int, Map<String, dynamic>, DateTime) handler) =>
      _onIncidentHandlers.add(handler);
  void onMessage(Function(int, String, DateTime) handler) =>
      _onMessageHandlers.add(handler);
  void onPartnerOnline(Function(bool) handler) =>
      _onPartnerOnlineHandlers.add(handler);
  void onPartnerTyping(Function() handler) =>
      _onPartnerTypingHandlers.add(handler);
  void onNewConversation(Function(int, String, String, DateTime) handler) =>
      _onNewConversationHandlers.add(handler);

  Future<void> disconnect() async {
    await _hub.stop();
  }
}

File: Web/ngMunicipaliGo/src/app/hub/ChatSignalRService.ts
import { Injectable, OnDestroy } from '@angular/core';
import * as signalR from '@microsoft/signalr';
import { BehaviorSubject } from 'rxjs';
import { ApiService } from '../services/api-service';

export interface Message {
  fromCitizenId: number;
  message: string;
  sentAt: Date;
  sharedIncident?: any;
}

@Injectable({ providedIn: 'root' })
export class ChatSignalRService implements OnDestroy {

  private hub: signalR.HubConnection;

  // Single source of truth: citizenId -> unread count
  // App reads totalUnread from this, ChatComponent clears per-conversation
  unreadMap$ = new BehaviorSubject<Map<number, number>>(new Map());
  private activePartnerId: number | null = null;
  messages$        = new BehaviorSubject<Message[]>([]);
  partnerOnline$   = new BehaviorSubject<boolean>(false);
  partnerTyping$   = new BehaviorSubject<boolean>(false);
  lastMessage$     = new BehaviorSubject<{ fromCitizenId: number; message: string; sentAt: Date } | null>(null);
  newConversation$ = new BehaviorSubject<{
    fromCitizenId: number;
    fromName: string;
    message: string;
    sentAt: Date;
  } | null>(null);

  private typingTimeout: any;

  constructor(private api: ApiService) {
    this.hub = new signalR.HubConnectionBuilder()
      .withUrl(`${this.api['serverUrl']}/hubs/chat`, {
        accessTokenFactory: () => localStorage.getItem('token') ?? ''
      })
      .withAutomaticReconnect()
      .build();

    // ── ReceiveMessage ───────────────────────────────────────────────────────
    this.hub.on('ReceiveMessage', (fromCitizenId: number, message: string, sentAt: string) => {
      const msg: Message = { fromCitizenId, message, sentAt: new Date(sentAt) };
      this.messages$.next([...this.messages$.value, msg]);
      this.lastMessage$.next(msg);
      this.partnerTyping$.next(false);
    });

    // ── ReceiveIncident ──────────────────────────────────────────────────────
    this.hub.on('ReceiveIncident', (fromCitizenId: number, incident: any, sentAt: string) => {
    const msg: Message = {
        fromCitizenId,
        message: '',
        sentAt: new Date(sentAt),
        sharedIncident: incident
    };
    this.messages$.next([...this.messages$.value, msg]);
    this.lastMessage$.next({ fromCitizenId, message: '', sentAt: new Date(sentAt) });
    });

    // ── NewConversationMessage ───────────────────────────────────────────────
    // This fires when the partner is NOT in the room (background notification)
    this.hub.on('NewConversationMessage', (fromCitizenId: number, fromName: string, message: string, sentAt: string) => {
    if (fromCitizenId !== this.activePartnerId) {
        this._addUnread(fromCitizenId);
    }
    this.newConversation$.next({ fromCitizenId, fromName, message, sentAt: new Date(sentAt) });
    });

    // ── PartnerTyping ────────────────────────────────────────────────────────
    this.hub.on('PartnerTyping', () => {
      this.partnerTyping$.next(true);
      clearTimeout(this.typingTimeout);
      this.typingTimeout = setTimeout(() => this.partnerTyping$.next(false), 3000);
    });

    // ── Online/Offline ───────────────────────────────────────────────────────
    this.hub.on('PartnerOnline',  () => this.partnerOnline$.next(true));
    this.hub.on('PartnerOffline', () => this.partnerOnline$.next(false));
  }

  async connect(): Promise<void> {
    if (this.hub.state === signalR.HubConnectionState.Disconnected) {
      await this.hub.start();
    }
  }
  
  setActiveConversation(citizenId: number | null) {
    this.activePartnerId = citizenId;
  }

  async openConversation(partnerUserId: number): Promise<void> {
    const history = await this.api.GetMessages(partnerUserId);
    this.messages$.next(history.map(m => ({
      fromCitizenId: m.fromCitizenId,
      message: m.message,
      sentAt: new Date(m.sentAt),
      sharedIncident: m.sharedIncident
    })));
    await this.hub.invoke('OpenConversation', partnerUserId);
  }

  async closeConversation(partnerUserId: number): Promise<void> {
    await this.hub.invoke('CloseConversation', partnerUserId);
    this.messages$.next([]);
    this.partnerOnline$.next(false);
    this.partnerTyping$.next(false);
  }

  // Called when user opens a conversation — removes it from unread map
  markRead(citizenId: number): void {
    const map = new Map(this.unreadMap$.value);
    map.delete(citizenId);
    this.unreadMap$.next(map);
  }

  // Internal only
  private _addUnread(citizenId: number): void {
    const map = new Map(this.unreadMap$.value);
    map.set(citizenId, (map.get(citizenId) ?? 0) + 1);
    this.unreadMap$.next(map);
  }

  async sendMessage(toUserId: number, message: string): Promise<void> {
    const msg: Message = { fromCitizenId: -1, message, sentAt: new Date() };
    this.messages$.next([...this.messages$.value, msg]);
    await this.hub.invoke('SendMessage', toUserId, message);
  }

  async sendIncident(toCitizenId: number, incidentId: number, incident: any): Promise<void> {
    const msg: Message = { fromCitizenId: -1, message: '', sentAt: new Date(), sharedIncident: incident };
    this.messages$.next([...this.messages$.value, msg]);
    await this.hub.invoke('SendIncident', toCitizenId, incidentId);
  }

  async sendTyping(toCitizenId: number): Promise<void> {
    if (this.hub.state === signalR.HubConnectionState.Connected)
      await this.hub.invoke('Typing', toCitizenId);
  }

  async disconnect(): Promise<void> {
    if (this.hub.state !== signalR.HubConnectionState.Disconnected)
      await this.hub.stop();
  }

  ngOnDestroy(): void {
    this.disconnect();
  }
}